### PR TITLE
Adding controller APIs to fetch brokers information

### PR DESCRIPTION
**Pull request by fx19880617**
## Description
Adding controller APIs to fetch brokers information
Per issue: https://github.com/apache/incubator-pinot/issues/5682

Sample output:
```
➜ curl localhost:9000/brokers | jq .
{
  "tenants": {
    "DefaultTenant": [
      "Broker_127.0.0.1_8000"
    ]
  },
  "tables": {
    "airlineStats": [
      "Broker_127.0.0.1_8000"
    ]
  }
}

```
```
➜ curl localhost:9000/brokers/tenants | jq .
{
  "DefaultTenant": [
    "Broker_127.0.0.1_8000"
  ]
}
```
```
➜ curl localhost:9000/brokers/tenants/DefaultTenant | jq .
[
  "Broker_127.0.0.1_8000"
]
```
```
➜ curl localhost:9000/brokers/tables | jq .
{
  "airlineStats": [
    "Broker_127.0.0.1_8000"
  ]
}
```
```
➜ curl "localhost:9000/brokers/tables/airlineStats?type=OFFLINE" | jq .
[
  "Broker_127.0.0.1_8000"
]
```
```
➜ curl "localhost:9000/brokers/tables/airlineStats?type=REALTIME" | jq .
[
  "Broker_127.0.0.1_8000"
]
```
```
➜ curl localhost:9000/brokers/tables/airlineStats | jq .
[
  "Broker_127.0.0.1_8000"
]
```
```
➜ curl localhost:9000/brokers/tables/airlineStats1 | jq .
{
  "code": 404,
  "error": "Table 'airlineStats1' not found."
}

```
## Release Notes
Adding Controller APIs to fetch brokers information:
```
// Return all the {state} brokers mapping of tenants/tables
GET /brokers?state={online|offline}

// Return all the {state} brokers mapping per tenants
GET /brokers/tenants?state={online|offline}

// Return {state} brokers for a given {tenant}, returns all brokers if state is not specified.
GET /brokers/tenants/{tenant}?state={online|offline}

// Return all the {state} brokers mapping per table
GET /brokers/tenants?state={online|offline}

// Return {state} brokers for a given {table}_{type}, returns all brokers if state is not specified.
GET /brokers/tables/{table}?type={REALTIME|OFFLINE}&?state={online|offline}
```

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    applyStateChanges(tableBrokers, state);
+    return ImmutableList.copyOf(tableBrokers);
+  }
+
+  private void applyStateChanges(Set<String> brokers, String state) {
+    if (state == null) {
+      return;
+    }
+    switch (state) {
+      case "ONLINE":
```
Comment (kishoreg) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java:124]: use the constant from statemodel

**Comment by mcvsubbu** [comment]
In order to be consistent with other APIs, I would take the raw table name, and a query parameter of table type in the API (instead of tableName_OFFLINE. etc.)

**Comment by mcvsubbu** [comment]
Also, we don't allow admin to specify different broker tenant names for realtime and offline tables. I would combine the two in the output as well. Keeps it consistent with the table config settings.
Otherwise, you run the risk of an admin trying to add broker tenant as "myTenant_REALTIME" or "myTenant_OFFLINE" in the table config.

**Comment by fx19880617** [comment]
> In order to be consistent with other APIs, I would take the raw table name, and a query parameter of table type in the API (instead of tableName_OFFLINE. etc.)

done

**Comment by fx19880617** [comment]
> Also, we don't allow admin to specify different broker tenant names for realtime and offline tables. I would combine the two in the output as well. Keeps it consistent with the table config settings.
> Otherwise, you run the risk of an admin trying to add broker tenant as "myTenant_REALTIME" or "myTenant_OFFLINE" in the table config.

Correct, broker tenant has no offline/realtime suffix also changed the behavior for tables to remove suffix.

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@Api(tags = Constants.BROKER_TAG)
+@Path("/")
+public class PinotBrokerRestletResource {
+  public static final Logger LOGGER = LoggerFactory.getLogger(PinotBrokerRestletResource.class);
+  private static final String TYPE_REALTIME = "_REALTIME";
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java:49]: these two definitions are not used anywhere right?

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java [hunk 3]
```diff
@@ -103,28 +110,35 @@
...
-    if (_pinotHelixResourceManager.hasOfflineTable(actualTableName)) {
-      tableBrokers.addAll(_pinotHelixResourceManager.getBrokerInstancesForTable(actualTableName, TableType.OFFLINE));
-    }
-    if (_pinotHelixResourceManager.hasRealtimeTable(actualTableName)) {
-      tableBrokers.addAll(_pinotHelixResourceManager.getBrokerInstancesForTable(actualTableName, TableType.REALTIME));
+    try {
+      List<String> tableNamesWithType = _pinotHelixResourceManager
+          .getExistingTableNamesWithType(tableName, Constants.validateTableType(tableTypeStr));
+      if (tableNamesWithType.isEmpty()) {
+        throw new ControllerApplicationException(LOGGER, String.format("Table [%s] not found.", tableName),
```
Comment (mcvsubbu) [inline pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java:119]: nit:
```suggestion
        throw new ControllerApplicationException(LOGGER, String.format("Table '%s' not found.", tableName),
```

File: pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java (hunk 2 above)
Comment (fx19880617) [reply pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotBrokerRestletResource.java]: removed

**Pull request merged**